### PR TITLE
Mobile client: Integrate backend (Basic)

### PR DESCRIPTION
**Pull request by Arthi-chaud**
- create API interface
- Make the provider responsible from most of the api calls
- Fix some bugs

### CODE REVIEW

**Review by 0Nom4D — APPROVED** [review]

**Comment by 0Nom4D** [comment]
### Note I forgot to write

As said to @Arthi-chaud, it might be more logic to put aeris_api.dart file inside a service folder instead of inside the model one:

* Api Service File isn't a model and belong to a service.

File: mobile/lib/src/providers/pipelines_provider.dart [hunk 1]
```diff
@@ -1,96 +1,90 @@
...
+    return _pipelineCollection.pipelines[index];
+  }
+
+  /// Sets sorting method for pipelines, does not execute sorting
+  void setSortingMethod(PipelineCollectionSort sortingMethod) {
+    _pipelineCollection.sortingMethod = sortingMethod;
+  }
+
+  /// Sets sorting method for pipelines, does not execute sorting
+  void setDisabledSplit(bool split) {
```
Comment (AnonymusRaccoon) [inline mobile/lib/src/providers/pipelines_provider.dart:82]: Use properties to do that, we don't want raw getter/setter. See [this](https://stackoverflow.com/questions/27683924/how-do-getters-and-setters-change-properties-in-dart) for more references.

File: mobile/lib/src/providers/pipelines_provider.dart [hunk 2]
```diff
@@ -1,96 +1,90 @@
...
+
+  /// returns the pipeline a t the given index in the collection.
+  /// Warning: if the pipeline is updated, no call will be made to api
+  Pipeline getPipelineAt(int index) {
+    return _pipelineCollection.pipelines[index];
+  }
+
+  /// Sets sorting method for pipelines, does not execute sorting
+  void setSortingMethod(PipelineCollectionSort sortingMethod) {
+    _pipelineCollection.sortingMethod = sortingMethod;
```
Comment (AnonymusRaccoon) [inline mobile/lib/src/providers/pipelines_provider.dart:78]: Shouldn't you resort the collection here?

File: mobile/lib/src/views/pipeline_detail_page.dart [hunk 3]
```diff
@@ -75,9 +68,8 @@ class _PipelineDetailPageState extends State<PipelineDetailPage> {
                         onToggle: (value) {
                           setState(() {
                             pipeline.enabled = !pipeline.enabled;
+                            GetIt.I<AerisAPI>().updatePipeline(pipeline);
```
Comment (AnonymusRaccoon) [inline mobile/lib/src/views/pipeline_detail_page.dart:71]: This won't work if the updatePipeline awaits before changing the state. You should not have unawaited calls in a state call.

File: mobile/lib/src/views/pipeline_detail_page.dart [hunk 4]
```diff
@@ -107,10 +99,11 @@ class _PipelineDetailPageState extends State<PipelineDetailPage> {
                 if (newreaction != Reaction.template()) {
                   setState(() {
                     pipeline.reactions.add(newreaction);
+                    GetIt.I<AerisAPI>().updatePipeline(pipeline);
```
Comment (AnonymusRaccoon) [inline mobile/lib/src/views/pipeline_detail_page.dart:102]: Same issue.

File: mobile/lib/src/views/pipeline_detail_page.dart [hunk 5]
```diff
@@ -156,10 +151,13 @@ class _PipelineDetailPageState extends State<PipelineDetailPage> {
...
                     action: reaction,
-                    then: () => setState(() {}),
-                    onDelete: () {
+                    then: () => setState(() {
+                      GetIt.I<AerisAPI>().updatePipeline(pipeline);
+                    }),
+                    onDelete: () => setState((){
                       pipeline.reactions.remove(reaction);
-                    },
+                      GetIt.I<AerisAPI>().updatePipeline(pipeline);
```
Comment (AnonymusRaccoon) [inline mobile/lib/src/views/pipeline_detail_page.dart:159]: Same issue.

File: mobile/lib/src/views/pipeline_detail_page.dart [hunk 6]
```diff
@@ -133,19 +125,22 @@ class _PipelineDetailPageState extends State<PipelineDetailPage> {
...
+                style: const TextStyle(fontWeight: FontWeight.w500)),
             ActionCard(
                 leading: pipeline.trigger.service.getLogo(logoSize: 50),
                 title: pipeline.trigger.name,
                 trailing: ActionCardPopupMenu(
                     deletable: false,
                     action: pipeline.trigger,
-                    then: () => setState(() {}))),
+                    then: () => setState(() {
+                          GetIt.I<AerisAPI>().updatePipeline(pipeline);
```
Comment (AnonymusRaccoon) [inline mobile/lib/src/views/pipeline_detail_page.dart:142]: Same issue.

File: mobile/lib/src/views/pipeline_detail_page.dart [hunk 7]
```diff
@@ -156,10 +151,13 @@ class _PipelineDetailPageState extends State<PipelineDetailPage> {
                   trailing: ActionCardPopupMenu(
                     deletable: reaction != pipeline.reactions.first,
                     action: reaction,
-                    then: () => setState(() {}),
-                    onDelete: () {
+                    then: () => setState(() {
+                      GetIt.I<AerisAPI>().updatePipeline(pipeline);
```
Comment (AnonymusRaccoon) [inline mobile/lib/src/views/pipeline_detail_page.dart:155]: Same issue.

File: mobile/lib/src/views/pipeline_detail_page.dart (hunk 4 above)
Comment (AnonymusRaccoon) [inline mobile/lib/src/views/pipeline_detail_page.dart:102]: Same issue.

File: mobile/lib/src/views/create_pipeline_page.dart [hunk 8]
```diff
@@ -110,13 +109,12 @@ class _CreatePipelinePageState extends State<CreatePipelinePage> {
...
                             text: AppLocalizations.of(context).addReaction,
                             onTap: () async {
-                              // TODO add to db
                               reactions.add(Reaction.template());
                               showAerisCardPage(
                                       context,
                                       (_) => SetupActionPage(
                                           action: reactions.last))
-                                  .then(setState(() {}));
+                                  .then((_) => setState(() {}));
```
Comment (AnonymusRaccoon) [inline mobile/lib/src/views/create_pipeline_page.dart:117]: I think you are doing something wrong.

File: mobile/lib/src/views/setup_action_page.dart [hunk 9]
```diff
@@ -43,7 +45,10 @@ class _SetupActionPageState extends State<SetupActionPage> {
       onChanged: (service) {
         setState(() {
           serviceState = service;
-          // TODO call api to get available actions
+          availableActions = [];
+          GetIt.I<AerisAPI>().getActionsFor(serviceState!, widget.action).then((actions) => setState(() {
```
Comment (AnonymusRaccoon) [inline mobile/lib/src/views/setup_action_page.dart:49]: You should probably not run a setState in a setState, but I think the then is the right way

Comment (AnonymusRaccoon) [inline mobile/lib/src/views/setup_action_page.dart:49]: You should probably not run a setState in a setState, but I think the then is the right way

File: mobile/lib/src/widgets/home_page_sort_menu.dart [hunk 10]
```diff
@@ -42,19 +43,17 @@ class HomePageSortMenu extends StatelessWidget {
...
         if (sortingMethod == "") {
-          collectionProvider.pipelineCollection.sortingSplitDisabled =
-              !collectionProvider.pipelineCollection.sortingSplitDisabled;
+          collectionProvider.setDisabledSplit(split);
         } else {
-          collectionProvider.pipelineCollection.sortingMethod =
-              sortingMethod as PipelineCollectionSort;
+          collectionProvider.setSortingMethod(sortingMethod as PipelineCollectionSort);
         }
         collectionProvider.sortPipelines();
```
Comment (AnonymusRaccoon) [inline mobile/lib/src/widgets/home_page_sort_menu.dart:58]: This should be done with the setter and not manually

File: mobile/lib/src/widgets/pipeline_card.dart [hunk 11]
```diff
@@ -33,7 +32,7 @@ class _PipelineCardState extends State<PipelineCard> {
           showAerisCardPage(
             context,
             (context) => PipelineDetailPage(pipeline: widget.pipeline),
-          );
+          ).then((_) => setState(() {}));
```
Comment (AnonymusRaccoon) [inline mobile/lib/src/widgets/pipeline_card.dart:35]: Weird

Comment (AnonymusRaccoon) [inline mobile/lib/src/widgets/pipeline_card.dart:35]: Weird

**Comment by Arthi-chaud** [comment]
The Comments @AnonymusRaccoon left were mainly because the name of the function 'updatePipeline" (now 'editPipeline') was unclear: the function takes a pipeline as parameters, and push its info to the API

File: mobile/lib/src/models/reaction.dart [hunk 12]
```diff
@@ -1,3 +1,5 @@
+// ignore_for_file: hash_and_equals
```
Comment (GitBluub) [inline mobile/lib/src/models/reaction.dart:1]: Warning ignore ?

File: mobile/lib/src/models/trigger.dart [hunk 13]
```diff
@@ -1,3 +1,5 @@
+// ignore_for_file: hash_and_equals
```
Comment (GitBluub) [inline mobile/lib/src/models/trigger.dart:1]: Warning ignore ?

File: mobile/lib/src/models/trigger.dart [hunk 14]
```diff
@@ -33,6 +35,7 @@ class Trigger extends aeris_action.Action {
       : super(service: const Service.twitter(), name: '', parameters: {});
 
   @override
+  // ignore: avoid_renaming_method_parameters
```
Comment (GitBluub) [inline mobile/lib/src/models/trigger.dart:38]: Do you need to ignore this warning too ?

**Review by AnonymusRaccoon — APPROVED** [review]

**Review by GitBluub — CHANGES_REQUESTED** [review]

**Review by Octopus773 — APPROVED** [review]

**Review by GitBluub — APPROVED** [review]

**Pull request merged**